### PR TITLE
rgw: Enable store specific implementation for user list

### PR DESCRIPTION
**Pull request by jjxsg**
## Problem Description
Function for user list api are called directly from `rgw_admin.cc` (in case of CLI) and from `rgw_user.cc` (in case of REST API). These functions are closely integrated with rados which makes implementation of user list complex for other stores.

## Solution
Move function calls for user list to backend store. 

## Implementation
A new function `list_users()` is introduced in backend store. In case of rados, it will call the `meta_list_keys_init()`, `meta_list_keys_next()` and `meta_list_keys_complete()` functions. 
Other stores can independently implement `list_users()` function without following implementation of above 3 functions. 

Signed-off-by: Jeet Jain <jeet.jain@seagate.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
  - [x] Code refactoring
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by andriytk** [comment]
I think we should be consistent with aws S3 API as much as possible. If in some case it's not possible (for some reason) - we should be consistent with rgw + rados backend.

**Comment by jjxsg** [comment]
jenkins retest this please

**Comment by jjxsg** [comment]
@dang @mattbenjamin, @cbodley please review this PR.

**Comment by jjxsg** [comment]
@dang @mattbenjamin, @cbodley please review this PR.

File: src/rgw/rgw_user.cc [hunk 1]
```diff
@@ -2141,52 +2141,39 @@ int RGWUser::list(const DoutPrefixProvider *dpp, RGWUserAdminOpState& op_state,
...
+  if (ret != -ENOENT) {
+    for (std::list<std::string>::iterator iter = users.begin(); iter != users.end(); ++iter){
       formatter->dump_string("key", *iter);
-        ++count;
-      }
     }
-  } while (truncated && left > 0);
-  // close user id list section
+  }
+  count = users.size();
```
Comment (ronen-fr) [inline src/rgw/rgw_user.cc:2166]: shouldn't it be moved to before the '}'?

File: src/rgw/rgw_user.cc [hunk 2]
```diff
@@ -2141,52 +2141,39 @@ int RGWUser::list(const DoutPrefixProvider *dpp, RGWUserAdminOpState& op_state,
...
-    } if (ret != -ENOENT) {
-      for (std::list<std::string>::iterator iter = keys.begin(); iter != keys.end(); ++iter) {
+
+  int ret = store->list_users(dpp, metadata_key, op_state.marker, op_state.max_entries, handle, &truncated, users);
+  if (ret < 0 && ret != -ENOENT) {
+    cerr << "ERROR: can't get key: " << cpp_strerror(-ret) << std::endl;
+    return ret;
+  }
+  if (ret != -ENOENT) {
+    for (std::list<std::string>::iterator iter = users.begin(); iter != users.end(); ++iter){
```
Comment (ronen-fr) [inline src/rgw/rgw_user.cc:2162]: 'auto' for iterators (or - better still - range-for with const auto*)

File: src/rgw/rgw_sal_rados.cc [hunk 3]
```diff
@@ -1302,6 +1302,37 @@ int RadosStore::meta_remove(const DoutPrefixProvider* dpp, std::string& metadata
...
+  if (max_entries > max) {
+      max_entries = max;
+  }
+  int ret = meta_list_keys_init(dpp, metadata_key, marker, &handle);
+  if (ret < 0){
+    return ret;
+  }
+  do {
+    std::list<std::string> keys;
+    left = ((max_entries > 0) ? max_entries - count : max);
```
Comment (ronen-fr) [inline src/rgw/rgw_sal_rados.cc:1330]: The 'int32_t - uint64_t' is not safe (albeit it won't actually wrap here).
And it will trigger a compiler warning.

**Comment by cbodley** [comment]
@dang the new `list_users()` API still has the `marker` and `max_entries` fields for pagination, so i don't see an issue there

my only concern here is related to multisite, which relies on having these generic interfaces to read/write/list metadata. there's still a lot of zipper design work left to do here - especially if our goal is to support metadata replication between different kinds of stores - but i worry that moving to specialized APIs for each type of metadata is taking us in the opposite direction and could make multisite harder to support

> These functions are closely integrated with rados which makes implementation of user list complex for other stores.

@jjxsg maybe you could summarize what part of the existing meta_list_keys_ APIs are coupled with rados? this new `list_users()` appears to have all of the same arguments

**Comment by jjxsg** [comment]
> The new API appears to be a one-shot, creating a list and returning that list.

No, that's not the case. There is no change in the actual API flow. I have just added the wrapper function which calls the functions to perform open(), get_next(), close(). 
The goal of this change is to move the `meta_list_keys_init()`, `meta_list_keys_next()` and `meta_list_keys_complete()` function calls to the backed store rather than keeping it in `rgw_admin.cc` or `rgw_user.cc` so that any store other than rados can independently implement the user list API. 
In case of rados, overall functionality remains the same.

**Comment by jjxsg** [comment]
> @jjxsg maybe you could summarize what part of the existing meta_list_keys_ APIs are coupled with rados? this new list_users() appear to have all of the same arguments

@cbodley I assumed that there are some design considerations with rados due to which the list user api is implemented in this way but that's not the case as @dang explained it is modeled after POSIX directory listing with with an open(), get_next(), close() type design. 

With this change, open(), get_next(), close() functions will be called from the backend store. This will make implementation easier for other stores. If a store already has similar api implemented they will simply call it from list_users() functions or if some other store have their own design and does not want to follow open(), get_next(), close() like design they can easily implement it without touching the non-SAL code.

**Comment by jjxsg** [comment]
Thanks for the response @dang @cbodley @ronen-fr. Please let me know if you have any further questions.

**Pull request closed without merging**